src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -135,10 +134,16 @@ KafkaSink::KafkaSink(
     , partition_cnt(topic->getPartitionCount())
     , one_message_per_row(kafka.produceOneMessagePerRow())
     , topic_refresh_interval_ms(kafka.topicRefreshIntervalMs())
+    , max_message_batch_size(context->getSettingsRef().max_insert_block_size.value)
```

Comment:
`max_insert_block_size` is not for bytes, but for row count

Reply:
Let's meter the row count instead of the bytes which could be more efficient and less computation check ?

Reply:
Okay, I will add a new setting instead. Using number of rows could be easier to code, however, it's not user-friendly. It's very hard to control the size by number of row.

Reply:
from my view, honoring  ingest row count is good enough and honoring bytes size introduces copies which is not ideal

Reply:
Even we change the code to limit by row, the logic actually won't change much. Because, the buffer still has its limit, it's still possible that a row or a few rows, after encoded, it will exceed that limit. Thus, we will still need that `markOffset` logic to tell it if it gets a complete data or not. So in fact, before this PR, there is actually a potential bug that it will produce weird messages if the above case happens.

Reply:
We can do both count and size limits at the same time, just like a normal batching strategy does. With the size limit, it ensures that a message won't be oversized, while with the row limit, one can tune it to reduce the memcpy operations.

Reply:
I like the row based since it is very intuitive than size

Reply:
Updated. I also added two metrics to external streams: `MessagesBySize` and `MessagesByRow`, to help users to observe the behavior, so that they can adjust when it needs to.